[SYCL][ESIMD] Use vload/vstore for simd object getter/setter.

kbobrovs opened the pull request:
DO NOT REVIEW FIRST TWO COMMITS (covered by https://github.com/intel/llvm/pull/1881)

Use special intrinsics to access simd objects in the private address space
to disable standard LLVM optimizations on them.

Author: Gang Chen <gang.y.chen@intel.com>

Signed-off-by: Konstantin S Bobrovsky <konstantin.s.bobrovsky@intel.com>

kbobrovs commented:
@cmc-rep, please pick the PR as the author of the code

againull commented on llvm/test/SYCLLowerIR/esimd_lower_ldst.ll:
```diff
@@ -0,0 +1,32 @@
+; Function Attrs: norecurse nounwind
+define dso_local spir_func void @_Z3foov() local_unnamed_addr #1 {
+; CHECK-LABEL: @_Z3foov(
+; CHECK-NEXT: [[TMP1:%.*]] = call <16 x i32> @llvm.genx.vload.v16i32.p0v16i32(<16 x i32>* getelementptr inbounds (%"class._ZTSN4sycl5intel3gpu4simdIiLi16EEE.sycl::intel::gpu::simd", %"class._ZTSN4sycl5intel3gpu4simdIiLi16EEE.sycl::intel::gpu::simd"* @vg, i64 0, i32 0))
```
Is it expected that vload is not lowered in this case? It is not obvious why only vstore is lowered no the normal store.

cmc-rep replied:
Yes, this is we are testing. vload/vstore associated with global variable with special attritube stays as vload/vstore.  vload/vstore not associated with those global variables is changed to generic load/store. This is explained in the code comment

againull replied:
This means that the lowering of the load is not tested, please add additional test to check lowering of the load.

cmc-rep replied:
load/store are processed in the same fashion. I added one case for load and one case for store. That should be sufficient

againull replied:
I don't see the test to check lowering of the load. How I should know if your code works correctly for the case when lowering of the load should happen?

againull replied:
I don't see a test for load lowering, how I could be sure that your code for load lowering is working?

againull commented on llvm/lib/SYCLLowerIR/LowerESIMDVLoadVStore.cpp:
```diff
@@ -0,0 +1,142 @@
+
+bool ESIMDLowerLoadStore::runOnFunction(Function &F) {
+  bool Changed = false;
+  Changed |= lowerLoadStore(F);
```
Is it possible to replace these 3 lines with
return lowerLoadStore(F);  ?

cmc-rep replied:
Why. It works and it is correct. Do we really need to pick on this detail?

againull replied:
Is this review only about functional correctness?

cmc-rep replied:
That is fine. I can change this one.

againull commented on llvm/lib/SYCLLowerIR/LowerESIMDVLoadVStore.cpp:
```diff
@@ -0,0 +1,142 @@
+#include "llvm/IR/Module.h"
+#include "llvm/InitializePasses.h"
+#include "llvm/Support/Debug.h"
+#include "llvm/Support/raw_ostream.h"
```
Probably some includes are not needed.

cmc-rep replied:
Which includes are not needed. Please be specific

againull replied:
Review process doesn't work this way.
I pointed to "#include "llvm/Support/raw_ostream.h" which probably was used during debugging the code but is not needed in the published code. And you should check if other headers are needed.

cmc-rep replied:
I will delete raw_ostream.h

againull commented on llvm/include/llvm/SYCLLowerIR/LowerESIMD.h:
```diff
@@ -34,6 +34,9 @@ class SYCLLowerESIMDPass : public PassInfoMixin<SYCLLowerESIMDPass> {
 FunctionPass *createSYCLLowerESIMDPass();
 void initializeSYCLLowerESIMDLegacyPassPass(PassRegistry &);
 
+FunctionPass *createESIMDLowerLoadStorePass();
```
Pass is written to work with old pass manager only. Please support it for the new pass manager, SYCLLowerESIMDPass or other pass can be used as an example.

againull commented on llvm/lib/SYCLLowerIR/LowerESIMDVLoadVStore.cpp:
```diff
@@ -0,0 +1,142 @@
+  // lower all vload/vstore into normal load/store.
+  std::vector<Instruction *> ToErase;
+  for (Instruction &Inst : instructions(F)) {
+    if (GenXIntrinsic::isVLoadStore(&Inst)) {
```
This loop body is one huge block which is hard to read.

Could you please use more early exits there? I think it will be much easier to read. For example:
```
if (!GenXIntrinsic::isVLoadStore(&Inst))
      continue;
```
and so on.

cmc-rep replied:
I think this is just your personal preference. Other people may prefer not to use continue

againull replied:
This is not my personal preference. Here is the link to the LLVM coding guidelines:
https://llvm.org/docs/CodingStandards.html#use-early-exits-and-continue-to-simplify-code

cmc-rep replied:
Coding style is just a guideline. Individual developer should have some freedom in applying it. In this case, for example, I do not like the negative sense in the condition when using condition. I prefer to use positive sense whenever possible for the condition test. Clearly, you read this code, and it did not confuse you.

againull replied:
Individual reviewers should have freedom in insisting to apply such guidelines (which are essentially a set of BKMs to write robust/correct/maintainable code) if they seem them suitable. I'm exercising that freedom.

cmc-rep replied:
I will resolve this so we can move on.

As someone new to this project and this process, I do have some different thoughts on the code review process. Let me quote a piece I found  on the internet to support my concern:
"There’s a imbalanced power dynamic in code review between reviewer and developer. This imbalance of power is by design – one of the functions of code review is to serve as a pedagogical tool – bringing developers up-to-speed on a project’s standards quickly. Power imbalances in code review, if unchecked and persistent:
•	Demotivate existing contributors
•	Repel new contributions
•	Have a chilling effect on contributions in review
These outcomes all result in:
•	Fewer new contributors
•	Fewer contributions from existing contributors
•	A lower diversity of contributors
•	An overall slower pace of change "

So, Is there a way to avoid the hazards of power imbalance?

againull replied:
Ok, I understand, but I am not the person who determines the review process for this project.
@pvchupin @olegmaslovatintel @bader @romanovvlad Could you please help with this question?

pvchupin replied:
This project follows LLVM coding guidelines and this is requirement.

cmc-rep replied:
I have used "continue"

againull replied:
Probably you forgot to upload the new changes.

cmc-rep replied:
Not sure what I did wrong. It should be up to date now:  https://github.com/kbobrovs/llvm/blob/esimd-ldst/llvm/lib/SYCLLowerIR/LowerESIMDVLoadVStore.cpp

kbobrovs commented:
@cmc-rep, please check once again if review comments are addressed.
@againull, can you please take another look and see if there are unaddressed comments.
I had to rebase and force-push to resolve conflicts.

cmc-rep commented:
I have addressed all the comments last week

kbobrovs commented:
> I have addressed all the comments last week

@againull, please take another look

kbobrovs commented:
Had to rebase after ESIMD lowering merge. @againull, can you please approve again.
